packages/caver-wallet/src/keyring/singleKeyring.js

```diff
@@ -0,0 +1,208 @@
+     * @type {PrivateKey}
+     */
+    get key() {
+        return this._key
+    }
+
+    set key(keyInput) {
+        if (keyInput === null) {
+            this._key = null
+            return
+        }
+        this._key = keyInput instanceof PrivateKey ? keyInput : new PrivateKey(keyInput)
+    }
+
+    /**
+     * returns public key string.
+     *
+     * @return {string}
+     */
+    getPublicKey() {
+        return this.key.getPublicKey()
+    }
+
+    /**
+     * returns a copied singleKeyring instance
+     *
+     * @return {SingleKeyring}
+     */
+    copy() {
+        return new SingleKeyring(this.address, this.key)
+    }
+
+    /**
+     * signs with transactionHash with a key and returns signature(s).
+     *
+     * @param {string} transactionHash The hash of transaction.
+     * @param {string|number} chainId The chainId specific to the network.
+     * @param {number} role A number indicating the role of the key. You can use `caver.wallet.keyring.role`.
+     * @param {number} [index] The index of the key to be used. If index is undefined, all private keys in keyring will be used.
+     * @return {Array.<string>|Array.<Array.<string>>}
+     */
+    sign(transactionHash, chainId, role, index) {
+        validateForSigning(transactionHash, chainId)
+
+        const key = this.getKeyByRole(role)
+
+        if (index !== undefined) {
+            validateIndexWithKeys(index, 1)
+            return key.sign(transactionHash, chainId)
+        }
+
+        return [key.sign(transactionHash, chainId)]
+    }
+
+    /**
+     * signs with hashed message and returns result object that includes `signatures`, `message` and `messageHash`
+     *
+     * @param {string} message The message string to sign.
+     * @param {number} role A number indicating the role of the key. You can use `caver.wallet.keyring.role`.
+     * @param {number} [index] The index of the key to be used.
+     * @return {object}
+     */
+    signMessage(message, role, index) {
+        if (role === undefined) throw new Error(`role should be defined for signMessage. Please use 'caver.wallet.keyring.role'.`)
+        const messageHash = utils.hashMessage(message)
+
+        const key = this.getKeyByRole(role)
+        const signatures = []
+        if (index !== undefined) {
+            validateIndexWithKeys(index, 1)
+        }
+
+        signatures.push(key.signMessage(messageHash))
+        return {
+            messageHash,
+            signatures,
+            message,
+        }
+    }
+
+    /**
+     * returns keys by role.If the key of the role passed as parameter is empty, the default key is returned.
+     *
+     * @param {number} role A number indicating the role of the key. You can use `caver.wallet.keyring.role`.
+     * @return {Array.<PrivateKey>}
+     */
+    getKeyByRole(role) {
+        if (role === undefined) throw new Error(`role should be defined.`)
+        if (role >= KEY_ROLE.roleLast) throw new Error(`Invalid role number: ${role}`)
```

Comment:
Don't we need to check negative indexes?

Reply:
I added validation to all getKeyByRole functions :) Can you refersh?